src/indexsets.jl

```diff
@@ -0,0 +1,141 @@
+"""
+    AbstractSparsityPattern
+
+Abstract supertype of all sparsity pattern representations.
+
+## Expected interface
+
+* `myempty(::Type{MyPattern})`: return an empty pattern
+* `seed(::Type{MyPattern}, i::Integer)`: return an pattern that only contains the given index `i`
+* `Base.show`
+
+And for compatibility with tracers:
+
+| Tracer               | Required accessor function                           |
+|:---------------------|:-----------------------------------------------------|
+| `ConnectivityTracer` | `inputs(p::MyPattern)`                               |
+| `GradientTracer`     | `gradient(p::MyPattern)`                             |
+| `HessianTracer`      | `gradient(p::MyPattern)` AND `hessian(p::MyPattern)` | 
+"""
+abstract type AbstractSparsityPattern end
+
+# Utilities on AbstractSet
+"""
+    myempty(S)
+
+Constructor for an empty set-like data structure of type `S`.
+"""
+myempty(::Type{S}) where {S<:AbstractSet} = S()
+
+"""
+    myempty(S)
+
+Constructor for a set-like data structure of type `S` that only contains the given index `i`.
+"""
+seed(::Type{S}, i::Integer) where {S<:AbstractSet} = S([i])
+
+""""
+    product(a::S{T}, b::S{T})::S{Tuple{T,T}}
+
+Inner product of set-like inputs `a` and `b`.
+"""
+product(a::AbstractSet{I}, b::AbstractSet{I}) where {I} = Set((i, j) for i in a, j in b)
+
+## First order
+abstract type AbstractFirstOrderPattern <: AbstractSparsityPattern end
+
+"""
+$(TYPEDEF)
+
+First order sparsity pattern represented by an index set of non-zero values.
+Represented by a set of integer indices.
+
+## Expected interface
+
+The passed set type `S` has to implement:
+* `SparseConnectivityTracer.myempty`
+* `SparseConnectivityTracer.seed`
+* `SparseConnectivityTracer.product`
+* `Base.union`
+* `Base.union!`
+* `Base.iterate`
+* `Base.collect`
+* `Base.show`
+
+All of these methods are already implemented for `AbstractSet`s from Julia Base.
```

Comment:
That is precisely the kind of thing which adds to the confusion. A pattern _is not_ an `AbstractSet`, it is _implemented with_ one or two `AbstractSet`s. So if you want to add such a sentence, it should be something like
```
All of these methods are already implemented for `AbstractSet`s from Julia Base, which the most common pattern subtypes rely on.
```

Reply:
> which is used by the most common pattern subtypes.

No, `AbstractSet`s are **exclusively** used in what you'd call `SimpleIndexSetPattern`. They are NOT used in anything else, e.g. what you'd call
* `SharedIndexSetPattern`
* `DictIndexSetPattern`
* `BitMaskPattern`

And even in `SimpleIndexSetPattern`, they are not enforced.

Reply:
```
All of these methods are already implemented for `AbstractSet`s
from Julia Base, which is used by the most common `SimpleIndexSetPattern` subtypes.
```
Maybe this is what you're suggesting?